program/src/big_vec.rs

```diff
@@ -0,0 +1,490 @@
+    pub fn push<T: Pack>(&mut self, element: T) -> Result<(), ProgramError> {
+        let mut vec_len_ref = &mut self.data[0..VEC_SIZE_BYTES];
+        let mut vec_len = u32::try_from_slice(vec_len_ref)?;
+
+        let start_index = VEC_SIZE_BYTES + vec_len as usize * T::LEN;
+        let end_index = start_index + T::LEN;
+
+        vec_len += 1;
+        vec_len.serialize(&mut vec_len_ref)?;
+
+        if self.data.len() < end_index {
+            return Err(ProgramError::AccountDataTooSmall);
+        }
+        let element_ref = &mut self.data[start_index..end_index];
+        element.pack_into_slice(element_ref);
+        Ok(())
+    }
+
+    /// Get an iterator for the type provided
+    pub fn iter<'vec, T: Pack>(&'vec self) -> Iter<'data, 'vec, T> {
+        Iter {
+            len: self.len() as usize,
+            current: 0,
+            current_index: VEC_SIZE_BYTES,
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Get a mutable iterator for the type provided
+    pub fn iter_mut<'vec, T: Pack>(&'vec mut self) -> IterMut<'data, 'vec, T> {
+        IterMut {
+            len: self.len() as usize,
+            current: 0,
+            current_index: VEC_SIZE_BYTES,
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Find matching data in the array
+    pub fn find<T: Pack>(&self, data: &[u8], predicate: fn(&[u8], &[u8]) -> bool) -> Option<&T> {
+        let len = self.len() as usize;
+        let mut current = 0;
+        let mut current_index = VEC_SIZE_BYTES;
+        while current != len {
+            let end_index = current_index + T::LEN;
+            let current_slice = &self.data[current_index..end_index];
+            if predicate(current_slice, data) {
+                return Some(unsafe { &*(current_slice.as_ptr() as *const T) });
+            }
+            current_index = end_index;
+            current += 1;
+        }
+        None
+    }
+
+    /// Find matching data in the array
+    pub fn find_mut<T: Pack>(
+        &mut self,
+        data: &[u8],
+        predicate: fn(&[u8], &[u8]) -> bool,
+    ) -> Option<&mut T> {
+        let len = self.len() as usize;
+        let mut current = 0;
+        let mut current_index = VEC_SIZE_BYTES;
+        while current != len {
+            let end_index = current_index + T::LEN;
+            let current_slice = &self.data[current_index..end_index];
+            if predicate(current_slice, data) {
+                return Some(unsafe { &mut *(current_slice.as_ptr() as *mut T) });
+            }
+            current_index = end_index;
+            current += 1;
+        }
+        None
+    }
+}
+
+/// Iterator wrapper over a BigVec
+pub struct Iter<'data, 'vec, T> {
+    len: usize,
+    current: usize,
+    current_index: usize,
+    inner: &'vec BigVec<'data>,
+    phantom: PhantomData<T>,
+}
+
+impl<'data, 'vec, T: Pack + 'data> Iterator for Iter<'data, 'vec, T> {
+    type Item = &'data T;
```

Comment:
I think the item type will need to be `T`, not `&T`, to be able to back the iterator by [`read_unaligned`](https://doc.rust-lang.org/std/ptr/fn.read_unaligned.html).

Reply:
That means you make a copy on each iteration step. This is bad. Remember that our protocol works on-chain and if some clients will not be able to run cli on M1 it is totally fine. We can't sacrifice for convenience in protocol

Reply:
One thing only matters - that unaligned code works on-chain.

Reply:
Another workaround is this. We can use unaligned code in protocol, but use aligned code on a client side. We can **implement is later** without updating protocol. But protocol should be with unaligned code due to performance issues.

Reply:
It could be or it could not. But it brings complexity anyway. I suggest to not bother and postpone aligned client code implementation to a next PR. Anyway, this code is in rust tradition of doing things. We implement small pieces of unsafe code and then use them. All rust  stdis like that.